### PR TITLE
Camipro Recognition And EPFL Authentication

### PR DESCRIPTION
**Pull request by AlexisAllmnd**
### 🌟 **PR Description** 🌟

This Pull Request (PR) introduces **significant updates and features** to the PocketTutor app, ensuring better functionality and a smoother user experience. Here's a breakdown of the changes made:

### 🚀 **New Features Added**
1. **Camipro Card Scanning with SCIPER Detection** 📸:
  - Introduced a **SciperScanButton** that uses the **CameraX** and **ML Kit** to detect SCIPER numbers on Camipro cards.
   - Supports live text recognition and provides clear UI feedback during scanning.
   - Handles permissions gracefully (e.g., prompts for camera access).

2. **EPFL Verification Workflow** 🛡️:
   - Added a **CertificationViewModel** and a new **EpflVerificationRepository** to manage SCIPER verification.
   - The **EpflVerificationCard** UI component allows users to verify their EPFL status:
     - 🟢 Displays success with auto-filled profile details.
     - 🔴 Shows errors for invalid or network issues.

3. **Profile Updates** ✍️:
   - Supports **auto-fill** of verified data (e.g., name, academic level) upon successful SCIPER verification.
   - Disabled manual input fields when verified, ensuring data integrity.

4. **Screenshots** 📸 : 
image
image

---

### 🛠️ **Enhancements**
1. **UI Improvements** 🎨:
   - Refined the **bottom navigation bar** for better visual clarity and accessibility.
   - Improved the **Completed Lessons** screen, fixing color problem with stars.

2. **Permission Handling** 🔐:
   - Add handling for camera permissions with user-friendly error messages.

---

### ✅ **Testing and Validation**
1. **Comprehensive Unit and UI Tests** 🧪:
   - Verified all states of the **SciperScanButton** (initial, scanning, detected).
   - Tested the **EpflVerificationCard** for success, error, and idle states.
   - Added edge case tests for SCIPER validation (e.g., invalid inputs, parsing errors).

2. **Code Coverage** 🧹:
   - Ensured all branches in the SCIPER parsing logic are covered, including academic levels and sections.

3. **Mock Testing** 🤖:
   - Used mocked responses to simulate network requests for SCIPER verification.

---

### 🔗 **Technical Details**
1. **Libraries Added**:
   - 📦 **CameraX** for live camera preview and processing.
   - 📦 **ML Kit Text Recognition** for detecting SCIPER numbers.
   - 📦 **Jsoup** for parsing HTML responses from the EPFL website.

2. **Dependency Updates**:
   - Updated `gradle/libs.versions.toml` to include new libraries:
     - `camera-core`, `camera-lifecycle`, `play-services-mlkit-text-recognition`, and more.

3. **Refactored Tests** 🧹:
   - Improved test structure for better maintainability.
   - Introduced reusable mock data and helper functions for testing.

---

### 💡 **Next Steps**
1. Integrate certification in the app to show certified user and possibility for uncertified tutor to do it in the app.

### CODE REVIEW

File: app/src/main/java/com/github/se/project/ui/components/AcademicSelector.kt [hunk 1]
```diff
@@ -3,53 +3,50 @@ package com.github.se.project.ui.components
...
+                    MaterialTheme.shapes.small)
                 .padding(16.dp)
-                .testTag("academicLevelDropdown"),
+                .testTag("academicLevelDropdown")
+                .alpha(if (enabled) 1f else 0.4f), // Visual feedback for disabled state
         style = MaterialTheme.typography.bodyLarge)
 
     DropdownMenu(
-        expanded = expandedDropdown.value,
+        expanded = expandedDropdown.value && enabled,
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/components/AcademicSelector.kt:39]: Now that we have the Camipro scan feature, it’s good that manual entry is disabled once the scan has already provided the input.

File: app/src/main/java/com/github/se/project/ui/components/AcademicSelector.kt [hunk 2]
```diff
@@ -3,53 +3,50 @@ package com.github.se.project.ui.components
...
         onDismissRequest = { expandedDropdown.value = false },
-        modifier = Modifier.fillMaxWidth().zIndex(1f),
-        properties = PopupProperties(focusable = true)) {
-          AcademicLevel.entries.forEach { a ->
+        modifier = Modifier.fillMaxWidth()) {
+          AcademicLevel.entries.forEach { s ->
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/components/AcademicSelector.kt:42]: not sure why you renamed that "a" variable to "s" but why not

File: app/src/main/java/com/github/se/project/ui/components/EpflVerificationCard.kt
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/components/EpflVerificationCard.kt:1]: the scan button looks really, really, really cool. That composable is very responsive, I am pretty impressed I have to admit. amazing work!!

File: app/src/main/java/com/github/se/project/ui/lesson/CompletedLesson.kt [hunk 3]
```diff
@@ -188,7 +189,7 @@ private fun DisplayReview(lesson: Lesson, student: Profile) {
                                   contentDescription = "Star",
                                   tint =
                                       if (index < rating.grade) MaterialTheme.colorScheme.primary
-                                      else MaterialTheme.colorScheme.secondary,
+                                      else Color.Gray.copy(alpha = 0.5f),
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/lesson/CompletedLesson.kt:192]: its indeed better to display the stars with grey colour.

File: app/src/test/java/com/github/se/project/model/certification/CertificationViewModelTest.kt
Comment (alexs-rufer) [inline app/src/test/java/com/github/se/project/model/certification/CertificationViewModelTest.kt:1]: great testing. kudos 👍

**Review by alexs-rufer — APPROVED** [review]
Wow, I’m genuinely impressed by the amount of effort you’ve put in! I tested the feature, and it’s super responsive, with a really good looking UI. It might be helpful to include a video in your PR to better showcase the quality of your work by the way :). I’ve suggested a few minor changes, but nothing major. That said, I do have a security concern: it seems like someone could potentially impersonate another person by stealing their SCIPER to gain certification. I understand that verifying this might not be easy and could add complexity for the user, but I think it’s an important question that we might want to discuss with the group.The PR is a bit long, but it’s totally understandable given the size and the quality of this new feature. Great job, and keep it up! 🚀

**Review by alexs-rufer — APPROVED** [review]
Reapproval since I already approved and you just did minor changes to pass the CI.
